[clang] Add value_type attr, use it to add noalias when pass-by-value.

fhahn opened the pull request:
This adds an attribute version of -fpass-by-value-is-noalias (added in a874d63344)

Still needs proper docs.

Attribute to be used by libc++ types: https://github.com/llvm/llvm-project/issues/90119

fhahn committed:
[clang] Add value_type attr, use it to add noalias when pass-by-value.

This adds an attribute version of -fpass-by-value-is-noalias (added in
a874d63344)

Still needs proper docs.

ldionne commented:
CC @philnik777

philnik777 commented:
This is really sweet. However, I feel like this is a property that needs to propagate through types. We can't guarantee that a `std::vector<int, my_nasty_allocator<int>>` doesn't escape the pointer. Similarly, I think we'd like
```c++
struct my_pair {
  int i;
  std::string str;
};
```
to have this property.

efriedma-quic commented:
> I feel like this is a property that needs to propagate through types

You mean, similar to the way trivial_abi works?  That makes sense.

efriedma-quic commented:
Also, we probably want to specify the interaction between trivial_abi and value_type here... should trivial_abi imply value_type?

fhahn commented:
Thanks, I'll hope to get back to this and make the suggested adjustments soonish

fhahn committed:
!fixup add attr docs.

fhahn committed:
!fixup update wording

fhahn commented:
Just rebased the PR and 
> > I feel like this is a property that needs to propagate through types
> 
> You mean, similar to the way trivial_abi works? That makes sense.

I am not sure about propagating. I added some documentation to make the spec of the attribute clearer. 

Specifically, it requires the objects of the type to be considered copied into new objects when passing to functions, i.e. code using the type isn't allowed to rely on previous pointers to the actual object e.g. via copy elision or NRVO.

Similarly, I think the guarantees are also different to `trivial_abi`, as there could be other pointers to the same object even if it is trivial via copy elision or NRVO. There's a thread where this problem in general has been discussed extensively a while ago https://lists.llvm.org/pipermail/cfe-dev/2020-July/066357.html

erichkeane commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -9242,3 +9242,15 @@ Declares that a function potentially allocates heap memory, and prevents any pot
+  let Category = DocCatDecl;
+  let Content = [{
+The ``value_type`` attribute can be used to mark user-defined types as 'value
+types'. When objects of value types are passed value to functions, the objects
```
```suggestion
types'. When objects of value types are passed by value to functions, the objects
```

erichkeane commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -9242,3 +9242,15 @@ Declares that a function potentially allocates heap memory, and prevents any pot
+are always considered to be formally copied into a new object. This means the
+argument itself must be the only value referencing the passed object. This
+allows the optimizer to assume that no other pointers may alias the object when
+it is passed indirectly to a function..
```
```suggestion
it is passed indirectly to a function.
```

erichkeane commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -9242,3 +9242,15 @@ Declares that a function potentially allocates heap memory, and prevents any pot
+The ``value_type`` attribute can be used to mark user-defined types as 'value
+types'. When objects of value types are passed value to functions, the objects
+are always considered to be formally copied into a new object. This means the
+argument itself must be the only value referencing the passed object. This
```
I am having a hard time figuring out why pass-by-value doesn't already imply this?  Pass-by-value means it was copied into the new object, right?  So I guess I have a hard time figuring out what we really mean here.

erichkeane replied:
Ok, I see what you mean now.  Yeah, that doesn't look like something that could be a property of the type so much as the uses itself.  So I again find myself wondering what property of the type we can annotate that makes it usable in the opt requested.

rjmccall replied:
If the argument satisfies the formal conditions for being passed in registers, the implementation is allowed to introduce a copy despite NRVO, which might be enough to make any escaped/self pointers formally invalid even if we don't actually copy. But yeah, NRVO might mean we're out of luck for an arbitrary type to which that doesn't apply.

We could probably talk to the C++ committee about this if we can show that we could do useful optimizations except for some contrived and dangerous code pattern like escaping the address of a copyable class value prior to returning it via NRVO.

fhahn replied:
@jyknight thanks for sharing the example! I made a few changes to make it runable: https://clang.godbolt.org/z/WrE8Ga6d9

What's interesting is that we get different results with Clang with and without optimizations (without optimizations, there's a copy due to `byval`, while with optimizations to function is inlined and the byval is gone). GCC seems to create copies both with and without optimizations and gives consistent results.

On AArch64, the argument isn't marked as `byval`, so we get the same results with and w/o optimizations. But it behaves differently to X86 w/o optimizations.

So overall it seems like there already are a number of inconsistencies when trying to capture the address of a pass-by-value argument across multiple dimensions; IIUC the program should return the same results for both GCC, Clang, with opts, w/o opts and different architectures.

The attribute itself is intended to forbid such uses of the type, but I agree it is not clear if it would be possible to use other for types in libc++.

erichkeane commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -1972,6 +1972,16 @@ def TrivialABI : InheritableAttr {
+  // CPlusPlus language option.
+  let Spellings = [Clang<"value_type", 0>];
+  let Subjects = SubjectList<[CXXRecord]>;
+  let Documentation = [TrivialABIDocs];
```
Why are we adding docs, but setting the documentation link here to something else?
